### ISSUE TITLE
This does not work with global binaries on Windows

### ISSUE THREAD

**Issue by ehmicky** [ehmicky/nve#14]
When running global binaries on Windows, the system `node` is used instead of the one specified by `nve`:

```
$ nve 8 myGlobalBinary
```

This is because global binaries on Windows are triggered through a [shim file](https://github.com/npm/cmd-shim/blob/master/index.js) (available both as `.cmd` file or as a Bash script). That shim file is in the same directory as the system `node` and it explicitly uses it:

```cmd
@ECHO off
SETLOCAL
CALL :find_dp0

IF EXIST "%dp0%\node.exe" (
  SET "_prog=%dp0%\node.exe"
) ELSE (
  SET "_prog=node"
  SET PATHEXT=%PATHEXT:;.JS;=;%
)

"%_prog%"  "%dp0%\node_modules\myGlobalBinary\build\src\bin.js" %*
ENDLOCAL
EXIT /b %errorlevel%
:find_dp0
SET dp0=%~dp0
EXIT /b
```

In the code above, `%dp0%\node.exe` explicitly refer to the `node.exe` in the same directory as the global binary. This bypasses `nve` which relies on the `PATH` environment variable.

**Comment by tunnckoCore** [comment]
Oh, how sweet. Love Windows, its shit and how it drives us crazy every single time :rofl:

**Comment by ehmicky** [comment]
Correct!

Other nice things making this bug more complex:
  - the shim file shown above is for the latest version of `npm`. Older versions have slightly different shim files.
  - yarn introduces an additional one-line shim file that points to that other shim file shown above.

I have a solution though that I am currently implementing.

**Comment by ehmicky** [comment]
And also: each global binary actually has three shim files used on Windows:
  - a Batch file executed when using `cmd.exe`
  - a Bash file executed when using Cygwin or MSYS2 / Git for Windows
  - a Powershell file executed when using Powershell

**Comment by tunnckoCore** [comment]
Beautiful. :rofl: 

I don't see how you can fix this :D Windows is never ending stream of bullshit.

Great research anyway! :100:

**Comment by papb** [comment]
Looks quite a mess... Good luck :)

**Comment by ehmicky** [comment]
An additional obstacle to mention: `nyc` [monkey patches `child_process.spawn()`](https://github.com/istanbuljs/spawn-wrap) to force using a specific `node` shim file that it uses for test coverage instrumentation. By doing so, it bypasses `nve` which relies on the `PATH` environment variable being respected.

The current code for `nve` has a fix for this, but it somehow conflicts with the fix for this issue. 

I have an idea on how to fix it though, and it seems to be working.

**Comment by ehmicky** [comment]
I have finished the bug fix. The final solution turns out to be quite complex, but it's working.

If you're interested about how this was solved, feel free to check the [branch](https://github.com/ehmicky/nvexeca/compare/global-binaries-windows), especially this summarizing comment:

```
// npm install global binaries on Unix with symlinks. But on Windows, it creates
// a small `*.cmd` file (https://github.com/npm/cmd-shim) that just forwards to
// `{executable} ...` where executable is guessed from the shabang (if any).
// This is because Windows cannot execute shabang files otherwise.
// It also does it with a Bash file and a `*.ps1` file (Powershell).
// However the shim files prioritize the global binaries directory over the
// `PATH` environment variable by looking for a sibling `node.exe`. But `nve`
// needs to use the `PATH` environment variable.
// We fix this by copying the shim file to a directory that we push in front of
// the `PATH`. This removes that prioritization since there are no more
// `node.exe`.
// The fix is npm-specific. Yarn does not have that issue. Although it also
// checks for a sibling `node.exe`, there is never such a file in practice.
```

However I need the following PR: https://github.com/sindresorhus/path-key/pull/6 to be merged and released first. `path-key` is a low-level dependency of `execa` and of this project, and it turns out it hard a subtle bug which makes the above fix not work on Travis CI.

**Comment by sindresorhus** [comment]
> However I need the following PR: sindresorhus/path-key#6 to be merged and released first.

✅

**Comment by ehmicky** [comment]
Alright I was going to celebrate finally finishing that bug when I realize: `npm` and `npx` global binaries use **slightly** [different shim files](https://github.com/npm/cli/blob/b829d62c98/bin/npm), so I need to use a different logic for those!

So I fixed this bug in [`8.0.2`](https://github.com/ehmicky/nve/releases/tag/8.0.2) for all Windows global binaries except `npm` and `npx`. Back to fixing those :/

**Comment by ehmicky** [comment]
This is now completely fixed in [`8.0.3`](https://github.com/ehmicky/nve/releases/tag/8.0.3) including `npm` and `npx` binaries.
